draft-ietf-quic-transport.md

```diff
@@ -2678,6 +2690,29 @@ advised to bundle as few streams as necessary in outgoing packets without losing
 transmission efficiency to underfilled packets.
 
 
+## Extension Frames (#extension-frames)
+
+During the handshake, the client may offer the server a list of extensions it
+supports as a part of transport parameters, and the server may reply with the
+list of extensions it has accepted.
+
+Every extension that defines a codepoint in the transport parameters extension
+list MUST have a well-defined and ordered list of new frames it provides (that
+list MAY be empty), as well as the rules for parsing said frames.  The extension
+frame codepoints are allocated by assigning the frame numbers in order starting
+from 0x80, where the order is defined first by the order of extensions as
+specified in the server handshake, and then by the order they are specified in
+the extension definition.  So, for instance, consider the situation where there
+are two extensions, A and B; extension A defines frames A1 and A2, in that
+order, and extension B defines only the frame B.  Then server accepting "A, B"
+results in assignment "0x80 = A1, 0x81=A2, 0x82 = B", whereas "B, A" would
+result in "0x80 = B, 0x81 = A1, 0x82= A2".
```

Comment:
As Christian says, it would make sense to have the client offer a list of the things that it supports (a list of extension points) and the server to indicate assignments (a list of frame type to extension point mappings).

That leaves Mike's concern.  We haven't committed to having a negotiation pattern where the client speaks first.  This would make that commitment.  If you want to avoid that, you could say that the mapping needs to be respected by the receiving peer.  That is, the client advertises a mapping of extension frame codepoint to frame octet and the server uses that for frames that it sends to the client.  The server might advertise a different mapping and the client would have to respect that too.  That's a tiny bit more complicated, but probably still feasible.

Finally, with either suggestion, you don't have to limit extension frames to the 0x80-0xff range.  You can use any codepoint.